src/Validation/ScannableURLsRestController.php

```diff
@@ -112,6 +122,19 @@ public function get_items_permissions_check( $request ) {
 	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
 	 */
 	public function get_items( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		// Allow query parameter to force a response to be served with Standard mode (AMP-first). This is used as
+		// part of Site Scanning in order to determine if the primary theme is suitable for serving AMP.
+		if ( ! amp_is_canonical() && $request->get_param( 'force_standard_mode' ) ) {
+			add_filter(
+				'option_' . AMP_Options_Manager::OPTION_NAME,
+				function ( $options ) {
+					$options[ Option::THEME_SUPPORT ] = AMP_Theme_Support::STANDARD_MODE_SLUG;
+
+					return $options;
+				}
+			);
+		}
+
```

Comment:
This is very much inspired by a [similar filter](https://github.com/ampproject/amp-wp/pull/6615/files#diff-bf37afb8930a2ddc44fe3b31adf43ce659d74bab4469a89253ecd9a23c4e06bcR257-R265) introduced in #6615.

I was trying to write a unit test for that but I couldn't get the `option` filter to actually override the template mode (even though it works as expected on the live site). Here's what I added to the `ScannableURLsRestControllerTest.php`:

```diff
b/tests/php/src/Validation/ScannableURLsRestControllerTest.php
--- a/tests/php/src/Validation/ScannableURLsRestControllerTest.php	(revision 2540a67282420840f4c6cf109dc960b3db589970)
+++ b/tests/php/src/Validation/ScannableURLsRestControllerTest.php	(date 1635969779679)
@@ -142,6 +142,14 @@
 				$this->assertNull( $scannable_url_entry['stale'] );
 			}
 		}
+
+		// Test `force_standard_mode` query parameter.
+		$request_with_forced_standard_mode  = new WP_REST_Request( 'GET', '/amp/v1/scannable-urls', [ 'force_standard_mode' => true ] );
+		$response_with_forced_standard_mode = rest_get_server()->dispatch( $request_with_forced_standard_mode );
+
+		$this->assertFalse( $response_with_forced_standard_mode->is_error() );
+		$scannable_urls_in_standard_mode = $response_with_forced_standard_mode->get_data();
+		$this->assertGreaterThan( 2, count( $scannable_urls_in_standard_mode ), 'Expected more than two URLs since in Standard mode.' );
 	}
 
 	/**

```

Reply:
It's probably because the `options` are cached in a member variable:

https://github.com/ampproject/amp-wp/blob/46e0b76b86888c0168533ce31151144f10e13a9f/src/Validation/ScannableURLProvider.php#L79-L87

I determined this was indeed the case, but there was also a secondary cause: when an option is not set in the DB, the `option_{$option_name}` filter does _not_ apply. Instead, the `default_option_{$option_name}` applies. In the context of the unit tests, the option is not set here, so that is also why it was failing.

As opposed to a filter, I was originally thinking that `ScannableURLProvider` could have a `set_options` method that would allow you to override the options used for scanning. However, I see that this will not fully work because `amp_is_post_supported()` does not consider the options stored in `ScannableURLProvider` but rather the options stored in the DB.

So it seems filters are the way to go, and to remove any caching in `ScannableURLProvider`.

Oh, and a third reason this doesn't work is the `[ 'force_standard_mode' => true ]` was being passed in as the `$attributes` and not as params. I believe it should have been done as follows:

```php
// Test `force_standard_mode` query parameter.
$request_with_forced_standard_mode = new WP_REST_Request( 'GET', '/amp/v1/scannable-urls' );
$request_with_forced_standard_mode->set_param( 'force_standard_mode', 'true' );
```

Reply:
I've addressed these concerns in bf243abc1fc97aca08b75430933de465af8af3fe and 4c8d9690e70ce98e29b0c6be29de901699f0d659.